include/rocksdb/cloud/cloud_env_options.h

```diff
@@ -173,8 +173,8 @@ class CloudEnvOptions {
   // If keep_local_log_files is false, this specifies what service to use
   // for storage of write-ahead log.
   LogType log_type;
-  // The implementation of the LogController for storage of the write-ahead log.
   std::shared_ptr<CloudLogController> cloud_log_controller;
+  std::shared_ptr<CloudStorageProvider> storage_provider;
```

Comment:
Also, I think comments on these 2 options would be good - since this is public API and we need to be clear what these options are.

Are these 2 options affected by `cloud_type`?

Reply:
This is kind of related to https://github.com/rockset/rocksdb-cloud/issues/80.  

Currently, CloudType has no bearing on anything and is only used in in "Dump/GetPrintableOptions".  This variable could be fairly safely removed from the public API I think.

For LogController, there is no "hard" relationship between it and the CloudType.  Some Controllers may only work with certain CloudTypes (Kinesis may require AWS) but others may work on multiple types (Kafka should work anywhere).

For StorageProvider, there is currently a 1-1 relationship between the CloudEnv implementation and the StorageProvider (AWS Env implies S3 Storage; GCP should imply Google Storage).  I do not know if this will always be true -- do you?  

I could envision there being more StorageProviders going forward.  It should be fairly simple to write one that is a "Backup" that copies files from one directory to another.  This would be useful for at least testing purposes.  HDFS could also be a StorageProvider that is independent of CloudEnv implementation.  At least some of these providers are likely independent of CloudEnv implementation.   What about MinIO -- where would it fall?

So, with one implementation of a StorageProvider and one of a CloudEnv, it is hard to say if there is always a 1-1 relationship between these.  LogController we have already proved that 1-1 relationship does not hold.  

I will add a comment for the storage_provider variable but not sure what else you want here.

Reply:
@mrambacher I see - thanks for the explanation on `cloud_type`. 

Yeah I'm thinking about the fact that we have `cloud_type` and `log_type`, and then we have `cloud_log_provider` and `storage_provider`. I feel like there is a redundancy here.

I think a simpler way would be to just have `cloud_type` and `log_type`. `cloud_log_provider` and `storage_provider` should just be implementation details (and should not be exposed to user). Something like:

```
class CloudEnvOptions {
  public:
       // Specify the type of cloud-service to use.
      CloudType cloud_type;

      // If keep_local_log_files is false, this specifies what service to use
      // for storage of write-ahead log.
      LogType log_type;

       // Other options
      ....
};
```

Then somewhere in `CloudEnvImpl` (probably as `protected` members), you'll have 
```
class CloudEnvImpl {
  protected:
    std::shared_ptr<CloudLogController> cloud_log_controller_;
    std::shared_ptr<CloudStorageProvider> storage_provider_;
};
```

In other words, `cloud_log_provider` and `storage_provider` should be created based on `cloud_type` and `log_type`, accordingly. If, in the future, we decide to have things like `HDFS` or `MinIO`, we can add more `cloud_type`. 

That being said, I'm not super familiar with `HDFS` or `MinIO`, so I'd like @igorcanadi or @dhruba to chime in here.

Reply:
Even though - I see your point here, `MinIO` and `HDFS` are not really cloud type, they're more storage provider.

Reply:
Actually, base RocksDB already supports `HDFS` (through `env_hdfs.h`). 

The way I think about `HDFS` or `MinIO` is that, these are filesystem in the sense of how unix thinks of filesystem. Conceptually, there is no local filesystem, and the "local" filesystem that RocksDB sees is actually `HDFS` or `MinIO`.

I think of S3 and GCS differently though - they are more of a "backup" system. That means, we still operate on local filesystem (through `base_env_`), and then syncing (or backing up) to the cloud.  This allows us to choose where to backup to (setting `dst_bucket` differently from `src_bucket`), or not backing up at all (ephemeral mode).

That means, if user were to choose a "backup" filesystem of `HDFS`, as opposed to `HDFS` as `base_env_`, then we can add a new `cloud_type`, something like `kHdfsOnAWS` and `kHdfsOnGCS`. I don't see us supporting this anytime soon though.

Reply:
I agree with you that we should not have both.  However, I would get rid of cloud_type and log_type rather than the log_controller and storage_provider:
- Because the types are enums, they are limited in what they can do and describe.  There is no way to add new ones without updating the base source code.  Then there are expectations about what you do and do not support and who supports/provides it, etc
- The enum types really describe explicit functionality and imply the interface they are implementing, as opposed to describing what one needs to do to implement one.  
- With enums, types that may be strictly internal or for testing must now be exposed publicly.  

I think you kind of also made the point indirectly that the interfaces need to be "public" if HDFS and MinIO are to become StorageProviders. 

Also note that a type implies a bunch of other configuration (like Kafka requires Kafka options).  There would be no good way of doing things like "controller.reset(NewKafkaController(KafkaOptions))" using just enums.  

I believe I have mostly removed the uses of these two enums from the source code.  They are left there for the moment until the public API is modified.  I could remove them now, but was asked previously to try to make all of the public API changes/removals in as few change sets as possible :)

I do not disagree that the types and these classes are redundant, just on which one should be removed.

Reply:
@mrambacher yeah you made a good point here - using enum would be very hard for configurations.

Reply:
Let's wait for @dhruba and @igorcanadi chime in, and let me think about this a little more

Reply:
I agree with you on the difference with the current HDFS implementation.  The "StorageProvider" is currently providing an "Object" file system where we are reading/writing entire objects rather than pieces.  Thinking of it as a backup also makes sense.  

And I am not suggesting implementing HDFS in this way in the short term, just that HDFS can also be used as a Object Store.

Now where it gets really interesting is if you want to use the Object Store as more than objects (think Athena or Presto)...